### PR TITLE
A0-1585: Replace AuthorityId with parameter

### PR DESCRIPTION
**Pull request by timorl**
# Description

Another step in making `validator_network` more independent. Will permit using a single `MockMultiaddress` everywhere, which will make testing simpler. For now it added another parameter to many structs, but at least for some this will not be permanent – we will just need to extract it from the `Multiaddress`, but this work is better left for after we make a `networking` module encompassing all this.

## Type of change

# Checklist:

- I have made corresponding changes to the existing documentation
- I have created new documentation

### CODE REVIEW

File: finality-aleph/src/tcp_network.rs [hunk 1]
```diff
@@ -66,6 +70,28 @@ impl Listener for TcpListener {
...
+    }
+}
+
+#[async_trait::async_trait]
+impl PrivateKey for AuthorityPen {
+    type Signature = Signature;
+    type PublicKey = AuthorityId;
+
+    async fn sign(&self, message: &[u8]) -> Self::Signature {
+        AuthorityPen::sign(self, message).await
```
Comment (maciejnems) [inline finality-aleph/src/tcp_network.rs:87]: Can't it be just `self.sign(message).await`?

**Review by maciejnems — CHANGES_REQUESTED** [review]

File: finality-aleph/src/tcp_network.rs (hunk 1 above)
Comment (timorl) [reply finality-aleph/src/tcp_network.rs:87]: [Ah, you fell for one of the classic blunders](https://www.youtube.com/watch?v=J3Z4yW60I2w)! I'll let you figure out why by yourself. :3

Comment (maciejnems) [reply finality-aleph/src/tcp_network.rs:87]: Ahhh, yes...

File: finality-aleph/src/validator_network/crypto.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+{
+    type Signature: Send + Sync + Clone + Codec;
+
+    /// Verify whether the message has been signed with the associated private key.
+    fn verify(&self, message: &[u8], signature: &Self::Signature) -> bool;
+}
+
+/// Private key for signing messages, with an associated public key.
+#[async_trait::async_trait]
+pub trait PrivateKey: Clone + Send + Sync + 'static {
```
[the remark this answers is not in the document]
Comment (timorl) [reply finality-aleph/src/validator_network/crypto.rs]: _Massively_ better, thank you.

File: finality-aleph/src/validator_network/manager/direction.rs [hunk 3]
```diff
@@ -3,16 +3,14 @@ use std::{
...
+use crate::validator_network::{Data, PublicKey};
 
 /// Data about peers we know and whether we should connect to them or they to us. For the former
 /// case also keeps the peers' addresses.
-pub struct DirectedPeers<A: Data> {
-    own_id: AuthorityId,
-    outgoing: HashMap<AuthorityId, Vec<A>>,
-    incoming: HashSet<AuthorityId>,
+pub struct DirectedPeers<PK: PublicKey, A: Data> {
+    own_id: PK,
```
[the remark this answers is not in the document]
Comment (timorl) [reply finality-aleph/src/validator_network/manager/direction.rs:11]: Ehh, here (within `manager`) it's treated exclusively as an ID, so I would leave this name. I removed some leftover `authority_id` though, and changed this in other parts of the code.

File: finality-aleph/src/validator_network/protocols/handshake.rs [hunk 4]
```diff
@@ -77,23 +73,28 @@ impl Challenge {
 /// Handshake response. Contains public key of the creator, and signature
 /// related to the received challenge.
 #[derive(Debug, Clone, Encode, Decode)]
-struct Response {
-    id: AuthorityId,
-    signature: Signature,
+struct Response<PK: PublicKey> {
+    id: PK,
```
[the remark this answers is not in the document]
Comment (timorl) [reply finality-aleph/src/validator_network/protocols/handshake.rs]: Changed here as well.

**Review by maciejnems — APPROVED** [review]
Looks great <3

**Review by lesniak43 — APPROVED** [review]
names make sense <3

**Pull request merged**